### ISSUE TITLE
Allow bitsunion as type parameters?

### ISSUE THREAD

**Issue by Keno** [JuliaLang/julia#33387]
This works:
```
julia> struct Bar{T}
       end

julia> struct FooWithoutUnion
           x::Int
       end

julia> Bar{FooWithoutUnion(1)}
Bar{FooWithoutUnion(1)}
```
this doesnt:
```
julia> struct FooWithUnion
          x::Union{Int, Nothing}
       end

julia> Bar{FooWithUnion(1)}
ERROR: TypeError: in Type, in parameter, expected Type, got FooWithUnion
Stacktrace:
 [1] top-level scope at REPL[20]:1
```

Now that we treat unions of bitstypes basically as good as bitstypes it seems to me that there is no reason to disallow the latter.

**Comment by quinnj** [comment]
Yeah, I feel like I remember discussing this briefly when we originally did the isbits Union optimizations and decided to punt. I think it makes sense.

**Comment by vtjnash** [comment]
I don't think we decided to punt so much as didn't want to make this implementation details part of the visible API. I also would generally discourage making structs with Union fields and/or thinking of them as isbitstypes. There's no good reason to allow this.

**Comment by Keno** [comment]
Well, I'd generally be in favor of extending the set of things you can put in a type parameter to be all things have egality comparison that doesn't depend on object identity anywhere, independent of whether or not we use pointers under the hood or not. As for no good reason to allow them, I needed something like this. I'm using `0` as a sentinel value instead now, which works fine, but isn't super general.

**Comment by JeffBezanson** [comment]
Yes, I think the requirement for type parameters is that they be immutable and have `===` do what you want (which perhaps just means everything they point to is also immutable?).

**Comment by vtjnash** [comment]
Yes, that sounds about right. I think the largest possible valid class is: all primitives types, and all immutable types for which all field \*values\* are also valid (or undef or Symbol). This is based on that assumption that the only requirement is for deepcopy(x)===x. The question is just whether this is desirable?

**Comment by JeffreySarnoff** [comment]
(a case in point)
For Julia to do as it intends with `Nothing` and `Missing`, it was _necessary_ to have the isbits Union optimization.  Having `Union{Nothing, BitsType}, Union{Missing, Int64, Float64}` be useful as field types and appropriate for use realizing a parametric type moves toward _sufficient_.
